Add code snapshot from internal repo

DMahoney0 opened the pull request.

DMahoney0 committed:
Add code snapshot from internal repo

mgwelch commented:
Duplicate files added EnumList.csv and SetupEnumList.csv

mgwelch commented:
Speaking of enum css files. Rather than loading the enum list via C# why didn't you just create a sql script with one Insert per record? That is what I had suggested.

mgwelch commented on README.md:
```diff
@@ -1,2 +1,159 @@
+
+## Quick Extract
+
+To run the Quick Extract app, run the command below in from the `Quick Extract` folder. The results for the last day are stored in the current directory in a new file named timeseries.csv.
```
Might want to be more explicit about the name of the folder. I don't think it's Quick Extract, it's HistoricalDataFetcher.QuickExtract. Why not even link to the actual folder in the rep?

dmassimini replied:
I changed this to show the full folder name as well as a link to the folder in the updated readme.md file.

mgwelch commented on README.md:
````diff
@@ -1,2 +1,159 @@
+To run the Quick Extract app, run the command below in from the `Quick Extract` folder. The results for the last day are stored in the current directory in a new file named timeseries.csv.
+
+```bash
+dotnet run --host <server.com> --u <Metasys Username> --p <Metasys Password>
````
in my experience `-` is used with short names and `--` is used with long names.

So this line I'd expect to be

```
dotnet run -h <server.com> -u <Metasys Username> -p <Metasys Password>
```

or

```
dotnet run --host <server.com> --user <Metasys Username> --password <Metasys Password>
```

mgwelch replied:
And generally an app supports both so you could mix and match.

dmassimini replied:
Our app does support both, these changes in the readme.md now reflect the "--user" and does not mention "--u".  The same is true for all other options as well.

mgwelch commented:
I'm getting runtime errors when using the quick start. For debugging purposes it might be nice to have .vscode/launch.json checked in.

Here's my error:

```bash
$ dotnet run --host hostname --u username --p password

Unhandled Exception: System.AggregateException: One or more errors occurred. (Invalid value '-244025.61' for parameter 'interval'.) ---> System.ArgumentException: Invalid value '-244025.61' for parameter 'interval'.
   at System.Timers.Timer..ctor(Double interval)
   at HistoricalDataFetcher.Classes.Utilities.ApiRequest.SetAuthenticationToken(AccessToken token) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:line 157
   at HistoricalDataFetcher.Classes.Utilities.ApiRequest.AuthenticateAsync() in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:line 136
   at HistoricalDataFetcher.Classes.Utilities.ApiRequest.InitializeAsync(ICache cache, String username, String password, String httpHost) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:line 58
   at HistoricalDataFetcher.Classes.Controller.Controller.InitApiRequestAsync(StartOptions startOptions) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Controller/Controller.cs:line 96
   at HistoricalDataFetcher.QuickExtract.Program.RunOptionsAsync(QuickStartOptions opts) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.QuickExtract/Program.cs:line 32
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at HistoricalDataFetcher.QuickExtract.Program.<>c.<Main>b__1_0(QuickStartOptions opts) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.QuickExtract/Program.cs:line 21
   at CommandLine.ParserResultExtensions.WithParsed[T](ParserResult`1 result, Action`1 action)
   at HistoricalDataFetcher.QuickExtract.Program.Main(String[] args) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.QuickExtract/Program.cs:line 21
```

mgwelch commented on HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:
```diff
@@ -0,0 +1,172 @@
+
+            using (var handler = new HttpClientHandler
+            {
+                ServerCertificateCustomValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true,
```
No, don't tell customers to do this.

DMahoney0 replied:
This has been addressed via the HandleCertificateError method

mgwelch commented on HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:
```diff
@@ -0,0 +1,172 @@
+            _username = username;
+            _password = password;
+            _httpHost = httpHost;
+            _urlBase = $"https://{_httpHost}/api";
```
why not set it to /api/v1?

yanli35 replied:
version can be specified in route or header, omitting it will default to version 1. if we put /v1 in the our sample code, we will be encouraging users to use route to specify version. is this what we want to promote?

mgwelch replied:
We want to encourage them to be explicit.

DMahoney0 replied:
This has been changed to v1

mgwelch commented on HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:
```diff
@@ -0,0 +1,172 @@
+        {
+            var body = $"{{\"username\": \"{_username}\",\"password\": \"{_password}\"}}";
+
+            var uri = new Uri($"{UrlBase}/api/authentication/login");
```
wrong api endpoint

mgwelch replied:
see documentation

yanli35 replied:
the external route wasn't ready when this is coded. will update.

mgwelch commented on HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:
```diff
@@ -0,0 +1,172 @@
+    
+            using (var handler = new HttpClientHandler
+            {
+                ServerCertificateCustomValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true
```
let's find all of these and fix them.

The only acceptable implementation is to examine the error, and if it's related to self signed certificate, AND if the customer passed a flag saying that they are okay with insecure mode, then you can return true. See how I did it in objectjs and alarmjs and nodekit and curl-scripts.

I allow for them to specify a cert saying they trust it, or they can pass --insecure and I warn them not to do that in product. We certainly don't want to hardcode a solution like this.

mgwelch replied:
However see my other comment on just telling a consumer how to setup their OS to trust their self-signed cert.

DMahoney0 replied:
This has been address via the HandleCertificateError method.

mgwelch commented on HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:
```diff
@@ -0,0 +1,172 @@
+            _accessTokenInstance = token;
+
+            //Have the interval expire the token 5 minutes before the token is set to expire to ensure there is some overlap
+            var interval = (_accessTokenInstance.Expires.AddMinutes(-5) - DateTime.Now).TotalMilliseconds;
```
There's something wrong with this line. I always get a negative number here.

mgwelch replied:
You'll want to guard against negative numbers and default to like 55 seconds in the case that the user has 1 minute expiration. (Which I think is what's happening here, because I setup a user with that)

Also, instead of using DateTime.Now, you'll want to use the Date on the headers returned form the server.

See what I did in nodekit: 

https://github.com/metasys-server/nodekit/blob/63ede3e1268ed4f82276506ac37e6acd2debfe8d/index.js#L175-L184

https://github.com/metasys-server/nodekit/blob/63ede3e1268ed4f82276506ac37e6acd2debfe8d/index.js#L157-L163

mgwelch replied:
I've confirmed that with a 30 minute expiration time this works okay. But if they have a 5 minute expiration this will likely fail especially if you are using DateTime.Now instead of server time. In my testing I could be almost a minute off from the server.

yanli35 replied:
the _accessTokenInstance.Expires is the expiration time returned from server, deserialized into local time, so we can use DateTime.Now to do the calculation.

server returns expiration date as a DateTimOffset object, which contains timezone info.

if the minimum expiration time in 1 minute, we can update this function to refresh 30 seconds before expiration date.

DMahoney0 replied:
Mike's suggestion has been implemented.

`var interval = (_accessTokenInstance.Expires.AddSeconds(-30) - DateTime.Now).TotalMilliseconds;`

mgwelch commented:
I get  warnings when running the quick extract. You should clean those up so a consumer doesn't think they did something wrong.

mgwelch commented:
Perhaps all of our code bases should just point to documentation like this: https://stackoverflow.com/a/623014/697188

And we just completely disallow insecure certs. Just tell consumers how to trust the certs.

mgwelch commented:
> Perhaps all of our code bases should just point to documentation like this: https://stackoverflow.com/a/623014/697188

Every OS has a way built in to trust certs.

mgwelch commented:
I noticed at least one sql file was detected as being binary: DBScripts/6.InsertAlarmData.sql

Might want to update .gitattributes to fix these.

DMahoney0 committed:
Fix invalid cert and access token object

Added an ‘InvalidCert” option to let user decide what to do when the
certificate received from server is a self signed certificate, when set
to true, the certificate error is ignored  and response is processed;
when set to false (default), the code looks at certificates on the
user computer to see if the certificate is stored locally, if it is,
the response is processed, if not, an exception will be thrown.

Changed login URL and the time interval for the access token refresh
rate. Changed the AccessToken object that was required to convert
the login credentials into the object.

DMahoney0 committed:
Update README, login errors, and exception logging

Updated the login error on incorrect host name, the ReadMe, and
added additional logging to try catch block exceptions

mgwelch commented on HistoricalDataFetcher.Classes/StartOptions/StartOptions.cs:
```diff
@@ -18,5 +18,8 @@ public class StartOptions
         [Option('f', "fqrs", HelpText = "The absolute path to the file containing the fullly qualified references")]
         public string FqrPath { get; set; }
+
+        [Option('i', "invalidcert", Default = false, HelpText = "Allow SSL certificate error in http response.")]
```
this could use some better warning. Not sure I'd know what that means

mgwelch replied:
You can just write up an issue

DMahoney0 replied:
Issue created: https://github.com/metasys-server/historian/issues/2

mgwelch commented:
I can't build one of the projects. It appears it's looking for a full framework 4.6.1:

```
HistoricalDataFetcher.Tests -> /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Tests/bin/Debug/netcoreapp2.1/HistoricalDataFetcher.Tests.dll
/usr/local/share/dotnet/sdk/2.1.302/Microsoft.Common.CurrentVersion.targets(1179,5): error MSB3644: The reference assemblies for framework ".NETFramework,Version=v4.6.1" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend. [/Users/cwelchmi/Documents/repos/metasys-server/historian/WindowsService/HistoricalDataFetcher.WindowsService.csproj]
```

dmassimini commented:
> I'm getting runtime errors when using the quick start. For debugging purposes it might be nice to have .vscode/launch.json checked in.
> 
> Here's my error:
> 
> ```shell
> $ dotnet run --host hostname --u username --p password
> 
> Unhandled Exception: System.AggregateException: One or more errors occurred. (Invalid value '-244025.61' for parameter 'interval'.) ---> System.ArgumentException: Invalid value '-244025.61' for parameter 'interval'.
>    at System.Timers.Timer..ctor(Double interval)
>    at HistoricalDataFetcher.Classes.Utilities.ApiRequest.SetAuthenticationToken(AccessToken token) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:line 157
>    at HistoricalDataFetcher.Classes.Utilities.ApiRequest.AuthenticateAsync() in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:line 136
>    at HistoricalDataFetcher.Classes.Utilities.ApiRequest.InitializeAsync(ICache cache, String username, String password, String httpHost) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs:line 58
>    at HistoricalDataFetcher.Classes.Controller.Controller.InitApiRequestAsync(StartOptions startOptions) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.Classes/Controller/Controller.cs:line 96
>    at HistoricalDataFetcher.QuickExtract.Program.RunOptionsAsync(QuickStartOptions opts) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.QuickExtract/Program.cs:line 32
>    --- End of inner exception stack trace ---
>    at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
>    at System.Threading.Tasks.Task.Wait()
>    at HistoricalDataFetcher.QuickExtract.Program.<>c.<Main>b__1_0(QuickStartOptions opts) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.QuickExtract/Program.cs:line 21
>    at CommandLine.ParserResultExtensions.WithParsed[T](ParserResult`1 result, Action`1 action)
>    at HistoricalDataFetcher.QuickExtract.Program.Main(String[] args) in /Users/cwelchmi/Documents/repos/metasys-server/historian/HistoricalDataFetcher.QuickExtract/Program.cs:line 21
> ```

Changes were made that now checks to see if the host provided is a valid URI location.  A similar check is made if the host exists.  In both cases the user will get a message letting them know the application could not login.

dmassimini commented:
> I get warnings when running the quick extract (during compilation step). You should clean those up so a consumer doesn't think they did something wrong.

I cleaned up all the warnings, they should not show up anymore.

dmassimini commented:
> Duplicate files added EnumList.csv and SetupEnumList.csv

Both of these files have a purpose, and they are in 2 different file locations.  The EnumList.csv is needed for the Quick Extract.  The application will use this file in place of calling the API's to gather the Enum set information.  The second file, SetupEnumList.csv, is used to load the Enum set into the Database.  There are checks in place to make sure that the data is not duplicated in the database and this file can be replaced by updated files in the future to load any new enum sets that are released.

dmassimini commented:
> Speaking of enum csv files. Rather than loading the enum list via C# why didn't you just create a sql script with one Insert per record? That is what I had suggested.

We spoke about this and loading into the system using bulk import would have been fine, but we planned on using this to allow the file to get replaced in the future with new Enum sets and our code will take care of the duplicates entries.

DMahoney0 committed:
Add check to refresh the access token in case of 1 minute expiration

Added in a check to refresh the access token at a minimum of every 50
seconds. This solves an issues where an user could have a token
expiration of one minute.
